translate/convert/l20n2po.py

```diff
@@ -0,0 +1,117 @@
+# along with this program; if not, see <http://www.gnu.org/licenses/>.
+
+
+"""Convert Mozilla .l20n files to Gettext PO localization files.
+
+See: http://docs.translatehouse.org/projects/translate-toolkit/en/latest/commands/l20n2po.html
+for examples and usage instructions.
+"""
+
+from translate.storage import po, l20n
+
+
+class l20n2po(object):
+    """convert a .l20n file to a .po file for handling the
+    translation."""
+
+    def __init__(self, blankmsgstr=False, duplicatestyle="msgctxt"):
+        self.blankmsgstr = blankmsgstr
+        self.duplicatestyle = duplicatestyle
+
+    def convert_l20nunit(self, unit):
+        po_unit = po.pounit(encoding="UTF-8")
+        po_unit.setid(unit.getid())
+        po_unit.addlocation(unit.getid())
+        po_unit.source = unit.value
+        po_unit.addnote(unit.comment, "developer")
+
+        return po_unit
+
+    def convert_store(self, l20n_store):
+        """converts a .l20n file to a .po file..."""
+        target_store = po.pofile()
+        targetheader = target_store.header()
+        targetheader.addnote("extracted from %s" % l20n_store.filename,
+                             "developer")
+        l20n_store.makeindex()
+        for l20nunit in l20n_store.units:
+            pounit = self.convert_l20nunit(l20nunit)
+            target_store.addunit(pounit)
+        target_store.removeduplicates(self.duplicatestyle)
+        return target_store
+
+    def merge_stores(self, origl20nfile, translatedl20nfile):
+        """converts two .l20n files to a .po file..."""
+        target_store = po.pofile()
+        targetheader = target_store.header()
+        targetheader.addnote(
+            "extracted from %s, %s" % (origl20nfile.filename,
+                                       translatedl20nfile.filename),
+            "developer"
+        )
+        translatedl20nfile.makeindex()
+        for l20nunit in origl20nfile.units:
+            pounit = self.convert_l20nunit(l20nunit)
+            target_store.addunit(pounit)
+            for location in pounit.getlocations():
+                # TODO: check if using of `id_index` fits here in all cases
+                if location in translatedl20nfile.id_index:
+                    l20nunit = translatedl20nfile.id_index[location]
+                    pounit.target = l20nunit.target
+        target_store.removeduplicates(self.duplicatestyle)
+        return target_store
+
+
+def convertl20n(inputfile, outputfile, templatefile,
+                pot=False, duplicatestyle="msgctxt"):
+    inputstore = l20n.l20nfile(inputfile)
+    convertor = l20n2po(blankmsgstr=pot, duplicatestyle=duplicatestyle)
+    if templatefile is None:
+        outputstore = convertor.convert_store(inputstore)
+    else:
+        templatestore = l20n.l20nfile(templatefile)
+        outputstore = convertor.merge_stores(templatestore, inputstore)
+    if outputstore.isempty():
+        return 0
+    outputstore.serialize(outputfile)
+    return 1
```

Comment:
I would refactor this is in a similar way to `po2l20n` converter, so it is easier to test, refactor and extract common behavior with other converters:

```python
class l20n2po(object):
    """convert a .l20n file to a .po file for handling the translation."""

    def __init__(self, input_file, output_file, template_file,
                 blank_msgstr=False, duplicate_style="msgctxt"):
        """Initialize converter."""
        self.blank_msgstr = blank_msgstr
        self.duplicate_style = duplicate_style

        self.output_file = output_file

        self.input_store = l20n.l20nfile(input_file)
        self.output_store = po.pofile()
        
        if template_file is None:
            self.template_store = None
        else:
            self.template_store = l20n.l20nfile(template_file)

    def convert_unit(self, unit):
        """Convert a l20n unit to a PO unit."""
        po_unit = po.pounit(encoding="UTF-8")
        po_unit.setid(unit.getid())
        po_unit.addlocation(unit.getid())
        po_unit.source = unit.value
        po_unit.addnote(unit.comment, "developer")

        return po_unit

    def convert_store(self):
        """Convert a l20n file to a PO file."""
        targetheader = self.output_store.header()
        targetheader.addnote("extracted from %s" % self.input_store.filename,
                             "developer")
        self.input_store.makeindex()

        for l20nunit in self.input_store.units:
            self.output_store.addunit(self.convert_unit(l20nunit))

        self.output_store.removeduplicates(self.duplicate_style)

    def merge_stores(self):
        """Convert two l20n files to a single PO file."""
        targetheader = self.output_store.header()
        targetheader.addnote(
            "extracted from %s, %s" % (self.template_store.filename,
                                       self.input_store.filename),
            "developer"
        )
        self.input_store.makeindex()

        for l20n_unit in self.template_store.units:
            po_unit = self.convert_unit(l20n_unit)
            self.output_store.addunit(po_unit)

            for location in po_unit.getlocations():
                # TODO: check if using of `id_index` fits here in all cases
                if location in self.input_store.id_index:
                    input_l20n_unit = self.input_store.id_index[location]
                    po_unit.target = input_l20n_unit.target

        self.output_store.removeduplicates(self.duplicate_style)

    def run(self):
        if self.template_store is None:
            self.convert_store()
        else:
            self.merge_stores()

        if self.output_store.isempty():
            return 0

        self.output_store.serialize(self.output_file)
        return 1


def convertl20n(inputfile, outputfile, templatefile, pot=False,
                duplicatestyle="msgctxt"):
    return l20n2po(inputfile, outputfile, templatefile, blank_msgstr=pot,
                   duplicate_style=duplicatestyle).run()
```